text/0000-colocated-tests.md

```diff
@@ -0,0 +1,210 @@
+To counter this conflict, co-located tests must end with `.test`, e.g.: `my-component.test.js`. Because having extra periods in the filename isn't allowed in framework objects, using `.test` should be fairly safe to search for when building the app and test bundles.
+
+More concisely:
+
+For the app/addon bundle,
+ - previously
+   - include: `{app,addon}/**/*.{js,hbs}`
+ - now
+   - include: `{app,addon}/**/*.{js,hbs}`
+   - exclude: `{app,addon}/**/*.test.{js,ts}`
+
+For the test bundle,
+ - previously
+   - include: `tests/**/*.{js}`
+ - now
+   - include: `tests/**/*.{js}`
+   - include: `{app,addon}/**/*.test.{js}`
+
+
+Like with component co-location, there will be different preferences with how to group files together. It may be undesirable to have a flat list of 20 routes, and then an additional flat list of 20 tests for those routes all in the same folder.
+The `.ember-cli` file provides the option to further nest components' class and template files as `{app,addon}/components/{component-name}/index.{js,hbs}` via the `componentStructure` property (set to `nested`).
+This same flexibility should be available for routes, controllers, etc -- but is out of scope for this RFC as other resolver implementations may be in-flight.
+Avoiding implementing `componentStructure` for the rest of the `app` directory does not have any impact on test co-location as the resolver is not involved with finding test file locations. For example, it would be valid to have `app/routes/{route-name}.js` and tests located at `app/routes/__tests__/{route-name}.test.js`.
+
+### Addons, in-repo Addons and Engines
+
+For normal addons, living as their own entire project, the lookup rules for apps will apply.
+For in-repo addons and engines, there is the possibility that the implementation of this feature would enable things not possible today.
+For example, if the "test finder" followed the glob: `**/*(\.test.js|-test.js)`, (excluding node\_modules, etc), that would include tests that are in in-repo addons and engines,
+which today, do not support their own tests -- the in-repo addon or engine still wouldn't support their own tests, exactly, as they would run as part of the parent
+app's test suite. This might be desired behavior as it would allow the in-repo addon and engine to finally co-locate their tests with the behavior that those units implement.
+
+
+### Reference: Where do the existing tests get migrated to?
+While tests in `{app,addon}/` could live anywhere, these are to be the default locations for the generators to use.
+
+#### Route
+
+previously:
+ - `{app,addon}/routes/{route-name}.js`
+ - `tests/unit/{route-name}-test.js`
+now:
+ - `{app,addon}/routes/{route-name}.js`
+ - `{app,addon}/routes/{route-name}.test.js`
+pods:
+ - `{pod-namespace}/{route-name}/route.js`
+ - `{pod-namespace}/{route-name}/.test.js`
```

Comment:
This raises a red flag. Won't this collide with controller tests?

```
❯ ember g route foo --pods
  create app/foo/route.js
  // therefore: app/foo/.test.js
❯ ember g controller foo --pods
  create app/foo/controller.js
  // therefore: app/foo/.test.js
```

This does not seem right?

Reply:
I've been under the impression that controller tests are rarely used and there wouldn't be a conflict.

_but_, the controller blueprint could totally check if there is a name collision and then output `app/foo/.controller.test.js